File: src/schemas/citation.py
"""Pydantic schemas for citation verification API.

This module defines all request/response schemas for the claim-level
citation verification feature, including:
- Claim responses with verification status
- Evidence span responses with source metadata
- Citation correction responses
- Numeric claim details with QA verification
- Verification summaries and metrics
- Provenance export formats
"""

from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import Field

from src.schemas.common import BaseSchema


# Enums for API layer (mirror database enums)
class ClaimTypeEnum(str, Enum):
    """Type of claim."""

    GENERAL = "general"
    NUMERIC = "numeric"


class VerificationVerdictEnum(str, Enum):
    """Four-tier verification verdict."""

    SUPPORTED = "supported"
    PARTIAL = "partial"
    UNSUPPORTED = "unsupported"
    CONTRADICTED = "contradicted"


class ConfidenceLevelEnum(str, Enum):
    """HaluGate-style confidence level."""

    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class CorrectionTypeEnum(str, Enum):
    """Citation correction type."""

    KEEP = "keep"
    REPLACE = "replace"
    REMOVE = "remove"
    ADD_ALTERNATE = "add_alternate"


class DerivationTypeEnum(str, Enum):
    """How a numeric value was obtained."""

    DIRECT = "direct"
    COMPUTED = "computed"


# Source metadata (denormalized for evidence cards)
class SourceMetadataResponse(BaseSchema):
    """Source metadata for evidence cards."""

    id: UUID
    title: str | None = None
    url: str | None = None
    author: str | None = None
    published_date: str | None = None
    content_type: str | None = None
    total_pages: int | None = None


# Evidence span response
class EvidenceSpanResponse(BaseSchema):
    """Evidence span from a source document."""

    id: UUID
    source_id: UUID
    quote_text: str
    start_offset: int | None = None
    end_offset: int | None = None
    section_heading: str | None = None
    relevance_score: float | None = None
    has_numeric_content: bool = False
    # Denormalized source metadata for convenience
    source: SourceMetadataResponse | None = None


# Citation response
class CitationResponse(BaseSchema):
    """Citation link between claim and evidence."""

    evidence_span: EvidenceSpanResponse
    confidence_score: float | None = None
    is_primary: bool = True


# Citation correction response
class CitationCorrectionResponse(BaseSchema):
    """Citation correction record."""

    id: UUID
    correction_type: CorrectionTypeEnum
    original_evidence: EvidenceSpanResponse | None = None
    corrected_evidence: EvidenceSpanResponse | None = None
    reasoning: str | None = None


# QA verification result for numeric claims
class QAVerificationResult(BaseSchema):
    """QA-based verification result for numeric claims."""

    question: str
    claim_answer: str
    evidence_answer: str
    match: bool
    normalized_comparison: dict[str, float] | None = None


# Computation step for derived values
class ComputationStep(BaseSchema):
    """Single computation step for derived numeric values."""

    operation: str
    inputs: list[dict] = Field(default_factory=list)
    result: float


# Numeric claim detail
class NumericClaimDetail(BaseSchema):
    """Extended details for numeric claims."""

    raw_value: str
    normalized_value: float | None = None
    unit: str | None = None
    entity_reference: str | None = None
    derivation_type: DerivationTypeEnum
    computation_details: dict | None = None
    assumptions: dict | None = None
    qa_verification: list[QAVerificationResult] | None = None


# Full claim response
class ClaimResponse(BaseSchema):
    """Full claim with verification and citation data."""

    id: UUID
    claim_text: str
    claim_type: ClaimTypeEnum
    confidence_level: ConfidenceLevelEnum | None = None
    position_start: int
    position_end: int
    verification_verdict: VerificationVerdictEnum | None = None
    verification_reasoning: str | None = None
    abstained: bool = False
    citations: list[CitationResponse] = Field(default_factory=list)
    corrections: list[CitationCorrectionResponse] = Field(default_factory=list)
    numeric_detail: NumericClaimDetail | None = None
    # Primary citation key for frontend mapping (e.g., "Arxiv", "Zhipu")
    citation_key: str | None = None
    # All citation keys in this claim for multi-marker sentences
    citation_keys: list[str] | None = None


# Verification summary
class VerificationSummary(BaseSchema):
    """Verification statistics for a message."""

    total_claims: int
    supported_count: int
    partial_count: int
    unsupported_count: int
    contradicted_count: int = 0
    abstained_count: int = 0
    unsupported_rate: float = Field(
        ge=0.0,
        le=1.0,
        description="Percentage of claims that are unsupported (0.0-1.0)",
    )
    contradicted_rate: float = Field(
        default=0.0,
        ge=0.0,
        le=1.0,
        description="Percentage of claims that are contradicted (0.0-1.0)",
    )
    warning: bool = Field(
        description="True if unsupported_rate > 0.20 or contradicted_rate > 0.05",
    )


# Correction metrics
class CorrectionMetrics(BaseSchema):
    """Citation correction statistics for a message."""

    total_corrections: int
    keep_count: int
    replace_count: int
    remove_count: int
    add_alternate_count: int
    correction_rate: float = Field(
        ge=0.0,
        le=1.0,
        description="Percentage of claims that needed correction (0.0-1.0)",
    )


# Response for GET /messages/{id}/claims
class MessageClaimsResponse(BaseSchema):
    """Response containing all claims for a message."""

    message_id: UUID
    claims: list[ClaimResponse]
    verification_summary: VerificationSummary
    correction_metrics: CorrectionMetrics | None = None


# Response for GET /claims/{id}/evidence
class ClaimEvidenceResponse(BaseSchema):
    """Response containing evidence for a specific claim."""

    claim_id: UUID
    claim_text: str
    verification_verdict: VerificationVerdictEnum | None = None
    citations: list[CitationResponse]


# Provenance export format
class ClaimProvenanceExport(BaseSchema):
    """Single claim in provenance export format."""

    claim_text: str
    claim_type: ClaimTypeEnum
    verdict: VerificationVerdictEnum | None = None
    confidence_level: ConfidenceLevelEnum | None = None
    citations: list[dict] = Field(default_factory=list)
    numeric_detail: NumericClaimDetail | None = None
    corrections: list[dict] = Field(default_factory=list)


class ProvenanceExport(BaseSchema):
    """Full provenance export for a message."""

    exported_at: datetime
    message_id: UUID
    claims: list[ClaimProvenanceExport]
    summary: VerificationSummary

File: databricks-deep-research-app/frontend/src/types/citation.ts
/**
 * Citation type definitions for claim-level attribution.
 *
 * This module defines TypeScript types for the 6-stage citation verification pipeline:
 * - Claims and their verification status
 * - Evidence spans and source metadata
 * - Citation links between claims and evidence
 * - Numeric claim details with QA verification
 * - Verification summaries and correction metrics
 */

/** Claim type classification */
export type ClaimType = 'general' | 'numeric';

/** Four-tier verification verdict */
export type VerificationVerdict =
  | 'supported'
  | 'partial'
  | 'unsupported'
  | 'contradicted';

/** HaluGate-style confidence level for routing */
export type ConfidenceLevel = 'high' | 'medium' | 'low';

/** Citation correction type */
export type CorrectionType = 'keep' | 'replace' | 'remove' | 'add_alternate';

/** Numeric value derivation type */
export type DerivationType = 'direct' | 'computed';

/** Source metadata for evidence cards */
export interface SourceMetadata {
  id: string;
  title: string | null;
  url: string | null;
  author: string | null;
  publishedDate: string | null;
  contentType: string | null;
  /** Total pages in document (for PDFs) */
  totalPages?: number;
}

/** Evidence span from a source document */
export interface EvidenceSpan {
  id: string;
  sourceId: string;
  quoteText: string;
  startOffset: number | null;
  endOffset: number | null;
  sectionHeading: string | null;
  relevanceScore: number | null;
  hasNumericContent: boolean;
  /** Denormalized source metadata for convenience */
  source: SourceMetadata;
}

/** Citation link between claim and evidence */
export interface Citation {
  evidenceSpan: EvidenceSpan;
  confidenceScore: number | null;
  isPrimary: boolean;
}

/** Citation correction record */
export interface CitationCorrection {
  id: string;
  correctionType: CorrectionType;
  originalEvidence: EvidenceSpan | null;
  correctedEvidence: EvidenceSpan | null;
  reasoning: string | null;
}

/** QA verification result for numeric claims */
export interface QAVerificationResult {
  question: string;
  claimAnswer: string;
  evidenceAnswer: string;
  match: boolean;
  normalizedComparison: {
    claimValue: number;
    evidenceValue: number;
  } | null;
}

/** Computation step for derived numeric values */
export interface ComputationStep {
  operation: string;
  inputs: Array<{
    value: number;
    sourceId: string;
    quote: string;
  }>;
  result: number;
}

/** Extended details for numeric claims */
export interface NumericClaimDetail {
  rawValue: string;
  normalizedValue: number | null;
  unit: string | null;
  entityReference: string | null;
  derivationType: DerivationType;
  computationDetails: {
    steps: ComputationStep[];
    formula: string | null;
  } | null;
  assumptions: {
    currencyYear: number | null;
    exchangeRate: number | null;
    roundingMethod: string | null;
    [key: string]: unknown;
  } | null;
  qaVerification: QAVerificationResult[] | null;
}

/** Full claim with all citation and verification data */
export interface Claim {
  id: string;
  claimText: string;
  claimType: ClaimType;
  confidenceLevel: ConfidenceLevel | null;
  positionStart: number;
  positionEnd: number;
  verificationVerdict: VerificationVerdict | null;
  verificationReasoning: string | null;
  abstained: boolean;
  citations: Citation[];
  corrections: CitationCorrection[];
  numericDetail: NumericClaimDetail | null;
  /** Primary citation key (e.g., "Arxiv", "Zhipu", "Github-2") */
  citationKey: string | null;
  /** All citation keys in this claim for multi-marker sentences */
  citationKeys: string[] | null;
}

/** Verification summary for a message */
export interface VerificationSummary {
  totalClaims: number;
  supportedCount: number;
  partialCount: number;
  unsupportedCount: number;
  contradictedCount: number;
  abstainedCount: number;
  unsupportedRate: number;
  contradictedRate: number;
  warning: boolean;
}

/** Correction metrics for a message */
export interface CorrectionMetrics {
  totalCorrections: number;
  keepCount: number;
  replaceCount: number;
  removeCount: number;
  addAlternateCount: number;
  correctionRate: number;
}

/** Response for GET /messages/{id}/claims endpoint */
export interface MessageClaimsResponse {
  messageId: string;
  claims: Claim[];
  verificationSummary: VerificationSummary;
  correctionMetrics: CorrectionMetrics | null;
}

/** Response for GET /claims/{id}/evidence endpoint */
export interface ClaimEvidenceResponse {
  claimId: string;
  claimText: string;
  verificationVerdict: VerificationVerdict | null;
  citations: Citation[];
}

/** Provenance export format */
export interface ProvenanceExport {
  exportedAt: string;
  messageId: string;
  claims: Array<{
    claimText: string;
    claimType: ClaimType;
    verdict: VerificationVerdict | null;
    confidenceLevel: ConfidenceLevel | null;
    citations: Array<{
      sourceUrl: string | null;
      sourceTitle: string | null;
      quote: string;
      isPrimary: boolean;
    }>;
    numericDetail: NumericClaimDetail | null;
    corrections: CitationCorrection[];
  }>;
  summary: VerificationSummary;
}

// SSE Event Types for real-time updates

/** Claim generated during interleaved synthesis */
export interface ClaimGeneratedEvent {
  eventType: 'claim_generated';
  timestamp: string;
  claimText: string;
  positionStart: number;
  positionEnd: number;
  evidencePreview: string;
  confidenceLevel: ConfidenceLevel;
}

/** Claim verification completed */
export interface ClaimVerifiedEvent {
  eventType: 'claim_verified';
  timestamp: string;
  claimId: string;
  claimText: string;
  positionStart: number;
  positionEnd: number;
  verdict: VerificationVerdict;
  confidenceLevel: ConfidenceLevel;
  evidencePreview: string;
  reasoning: string | null;
  /** Primary citation key for citationData mapping (e.g., "Arxiv", "Zhipu") */
  citationKey: string | null;
  /** All citation keys for multi-source claims */
  citationKeys: string[] | null;
}

/** Citation corrected during post-processing */
export interface CitationCorrectedEvent {
  eventType: 'citation_corrected';
  timestamp: string;
  claimId: string;
  correctionType: CorrectionType;
  reasoning: string | null;
}

/** Numeric claim detected */
export interface NumericClaimDetectedEvent {
  eventType: 'numeric_claim_detected';
  timestamp: string;
  claimId: string;
  rawValue: string;
  normalizedValue: string | null;
  unit: string | null;
  derivationType: DerivationType;
  qaVerified: boolean;
}

/** Verification summary for a message */
export interface VerificationSummaryEvent {
  eventType: 'verification_summary';
  timestamp: string;
  messageId: string;
  totalClaims: number;
  supported: number;
  partial: number;
  unsupported: number;
  contradicted: number;
  abstainedCount: number;
  citationCorrections: number;
  warning: boolean;
}

/** Union type for all citation-related SSE events */
export type CitationStreamEvent =
  | ClaimGeneratedEvent
  | ClaimVerifiedEvent
  | CitationCorrectedEvent
  | NumericClaimDetectedEvent
  | VerificationSummaryEvent;

/** Verdict color mapping for UI */
export const VERDICT_COLORS: Record<VerificationVerdict, string> = {
  supported: 'green',
  partial: 'amber',
  unsupported: 'red',
  contradicted: 'purple',
};

/** Verdict labels for display */
export const VERDICT_LABELS: Record<VerificationVerdict, string> = {
  supported: 'Supported',
  partial: 'Partially Supported',
  unsupported: 'Unsupported',
  contradicted: 'Contradicted',
};

/** Confidence level colors for UI */
export const CONFIDENCE_COLORS: Record<ConfidenceLevel, string> = {
  high: 'green',
  medium: 'amber',
  low: 'red',
};
